### PR TITLE
Switch vector and angle types to use userdata types

### PR DESCRIPTION
**Pull request by dvdvideo1234**
https://github.com/wiremod/wire/pull/2399

### CODE REVIEW

File: lua/entities/gmod_wire_expression2/core/custom/trackasmlib_wire.lua [hunk 1]
```diff
@@ -88,30 +88,23 @@ e2function array entity:trackasmlibSnapEntity(vector trHitPos  , string hdModel
...
 end
 
 __e2setcost(80)
 e2function array trackasmlibSnapNormal(vector ucsPos, angle  ucsAng   , string hdModel,
                                        number hdPoID, vector ucsOffPos, angle ucsOffAng)
   if(not enFlag) then return {} end
-  local stSpawn = asmlib.GetNormalSpawn(self.player,
-                                        asmlib.ToVector(ucsPos, wvX, wvY, wvZ),
-                                        asmlib.ToAngle(ucsAng, waP, waY, waR),
-                                        ucsAng,hdModel,hdPoID,
```
Comment (dvdvideo1234) [inline lua/entities/gmod_wire_expression2/core/custom/trackasmlib_wire.lua:108]: This has one `ucsAng` argument more

File: lua/trackassembly/trackasmlib.lua [hunk 2]
```diff
@@ -826,12 +826,6 @@ function GetColor(xR, xG, xB, xA)
   return Color(nR, nG, nB, nA)
 end
 
-function ToColor(vBase, pX, pY, pZ, vA)
```
Comment (dvdvideo1234) [inline lua/trackassembly/trackasmlib.lua]: Not used anywhere and not needed anymore

File: lua/trackassembly/trackasmlib.lua [hunk 3]
```diff
@@ -857,18 +851,6 @@ end
 
 ------------- ANGLE ---------------
 
-function ToAngle(aBase, pP, pY, pR)
```
Comment (dvdvideo1234) [inline lua/trackassembly/trackasmlib.lua:858]: Not used anywhere and not needed anymore

File: lua/trackassembly/trackasmlib.lua [hunk 4]
```diff
@@ -857,18 +851,6 @@ end
...
-  if(not aBase) then LogInstance("Base invalid"); return nil end
-  local aP, aY, aR = UseIndexes(pP, pY, pR, caP, caY, caR)
-  return Angle((tonumber(aBase[aP]) or 0), (tonumber(aBase[aY]) or 0), (tonumber(aBase[aR]) or 0))
-end
-
-function ExpAngle(aBase, pP, pY, pR)
```
Comment (dvdvideo1234) [inline lua/trackassembly/trackasmlib.lua]: Not used anywhere and not needed anymore: https://wiki.facepunch.com/gmod/Vector:SetUnpacked

File: lua/trackassembly/trackasmlib.lua [hunk 5]
```diff
@@ -899,18 +881,6 @@ end
 
 ------------- VECTOR ---------------
 
-function ToVector(vBase, pX, pY, pZ)
```
Comment (dvdvideo1234) [inline lua/trackassembly/trackasmlib.lua:894]: Not used anywhere and not needed anymore

File: lua/trackassembly/trackasmlib.lua [hunk 6]
```diff
@@ -899,18 +881,6 @@ end
...
-  if(not vBase) then LogInstance("Base invalid"); return nil end
-  local vX, vY, vZ = UseIndexes(pX, pY, pZ, cvX, cvY, cvZ)
-  return Vector((tonumber(vBase[vX]) or 0), (tonumber(vBase[vY]) or 0), (tonumber(vBase[vZ]) or 0))
-end
-
-function ExpVector(vBase, pX, pY, pZ)
```
Comment (dvdvideo1234) [inline lua/trackassembly/trackasmlib.lua]: Not used anywhere and not needed anymore

File: lua/trackassembly/trackasmlib.lua [hunk 7]
```diff
@@ -945,14 +915,6 @@ function SubVectorXYZ(vBase, nX, nY, nZ)
   return vBase
 end
 
-function NegVector(vBase, bX, bY, bZ)
```
Comment (dvdvideo1234) [inline lua/trackassembly/trackasmlib.lua]: Not used anywhere and not needed anymore: https://wiki.facepunch.com/gmod/Vector:Negate

File: lua/trackassembly/trackasmlib.lua [hunk 8]
```diff
@@ -1948,7 +1910,7 @@ function SetCenter(oEnt, vPos, aAng, nX, nY, nZ) -- Set the ENT's Angles first!
     LogInstance("Entity Invalid"); return Vector(0,0,0) end
   oEnt:SetPos(vPos); oEnt:SetAngles(aAng)
   local vCen, vMin = oEnt:OBBCenter(), oEnt:OBBMins()
-  NegVector(vCen); vCen[cvZ] = 0 -- Adjust only X and Y
+  vCen:Negate(); vCen[cvZ] = 0 -- Adjust only X and Y
```
Comment (dvdvideo1234) [inline lua/trackassembly/trackasmlib.lua]: https://wiki.facepunch.com/gmod/Vector:Negate

File: lua/weapons/gmod_tool/stools/trackassembly.lua [hunk 9]
```diff
@@ -464,7 +464,7 @@ function TOOL:IntersectSnap(trEnt, vHit, stSpawn, bMute)
     else asmlib.Notify(ply, "Model intersection mismatch !", "ERROR")
       asmlib.LogInstance("Model ray mismatch",gtLogs); return nil end
   end
-  local aOrg, vx, vy, vz = stSpawn.OAng, asmlib.ExpVector(stSpawn.PNxt)
+  local aOrg, vx, vy, vz = stSpawn.OAng, stSpawn.PNxt:Unpack()
```
Comment (dvdvideo1234) [inline lua/weapons/gmod_tool/stools/trackassembly.lua]: https://wiki.facepunch.com/gmod/Vector:Unpack

**Comment by SupinePandora43** [comment]
You could've put a wiremod PR link in description, so it would be properly linked.
Also It's better to have an informative PR name.

**Comment by dvdvideo1234** [comment]
@SupinePandora43 

You got a point there ;) I thought it was a better idea at the time and forgot about it lol :P

File: lua/entities/gmod_wire_expression2/core/custom/trackasmlib_wire.lua [hunk 10]
```diff
@@ -41,18 +41,18 @@ local gsCbcHash = "_wire" -- This keeps suffix realted to the file
 
 gsVarName = asmlib.GetAsmConvar("enwiremod", "NAM")
 cvarsRemoveChangeCallback(gsVarName, gsVarName..gsCbcHash)
-cvarsAddChangeCallback(gsVarName, function(sVar, vOld, vNew)
-  enFlag = ((tonumber(vNew) or 0) ~= 0) end, gsVarName..gsCbcHash)
+cvarsAddChangeCallback(gsVarName, function(sV, vO, vN)
```
Comment (dvdvideo1234) [inline lua/entities/gmod_wire_expression2/core/custom/trackasmlib_wire.lua]: Change locals here to some shorter names

**Pull request merged**